In [ ]:
!curl "https://raw.githubusercontent.com/pasricha/Subtask-A/master/Data.Assignment2.SemEvalTask9SubtaskA.csv" > Data.Assignment2.SemEvalTask9SubtaskA.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  729k  100  729k    0     0  3663k      0 --:--:-- --:--:-- --:--:-- 3645k


In [ ]:
#Importing the libraries
import numpy as np #Numpy is used for providig large, multi dimensional arrays and matrices. It also provides mathematical functions.
import pandas as pd #Pandas is used for analysis and manipulation of data stored in numerical tables and time series data structures.
# Read the CSV file.
df = pd.read_csv('Data.Assignment2.SemEvalTask9SubtaskA.csv', 
                 names=['id', 'text', 'label'], header=0)

# Set seed for reproducibility and shuffle the rows.
np.random.seed(888) #It is used to generate psuedo random numbers for random processes.
df = df.sample(frac=1).reset_index(drop=True)

# Store the data as a list of tuples where the first item is the text
# and the second item is the label.
data = [(text, label) for (idx, text, label) in df.values.tolist()]

# Create training and test sets.
train_texts, train_labels = [], []
test_texts, test_labels = [], []

#################### EDIT BELOW THIS LINE #########################
#Initialising the lists as per the given question
train_texts, train_labels = df.text[:5100],df.label[:5100]
test_texts, test_labels = df.text[5100:],df.label[5100:]



#################### EDIT ABOVE THIS LINE #########################

# Check that training set and test set are of the right size.
assert len(test_texts) == len(test_labels) == 1000
assert len(train_texts) == len(train_labels) == 5100

In [ ]:
import numpy as np
import pandas as pd

# Read the CSV file.
df = pd.read_csv('Data.Assignment2.SemEvalTask9SubtaskA.csv', 
                 names=['id', 'text', 'label'],header = 0 )

# Set seed for reproducibility and shuffle the rows.
np.random.seed(888)
df = df.sample(frac=1).reset_index(drop=True)

# Store the data as a list of tuples where the first item is the text
# and the second item is the label.
data = [(text, label) for (idx, text, label) in df.values.tolist()]

# Create training and test sets.
train_texts, train_labels = [], []
test_texts, test_labels = [], []

#################### EDIT BELOW THIS LINE #########################
train_texts, train_labels = df.text[:5100],df.label[:5100]
test_texts, test_labels = df.text[5100:],df.label[5100:]



#################### EDIT ABOVE THIS LINE #########################

# Check that training set and test set are of the right size.
assert len(test_texts) == len(test_labels) == 1000
assert len(train_texts) == len(train_labels) == 5100

In [ ]:
#Checking the length of the training dataset
len(train_texts ) 

5100

In [ ]:
#Checking the length of the training dataset
len(test_texts) #Checking the length of the testing dataset

1000

In [ ]:
print(df) 


          id                                               text  label
0       4191  "Warning 4668 is an extra one we have enabled ...      0
1     978_10  "Make it possible to optionally declare a func...      1
2     1274_2  "It has a new feature to add a route that will...      0
3      793_8  "It's nice to have a notification from a news ...      0
4      0_421  "And please don't always use security risk as ...      1
...      ...                                                ...    ...
6095    2833  "When the user will press button for next item...      0
6096  0_1000  "We end up writing lot of custom Styles in XAM...      0
6097    1254  "Allow UWP applications to tick a box in the ....      1
6098    8378   Allow JSON and/or XML to pass data between apps.      1
6099    1102  "I have added this whole ASN (AS8075 https://i...      0

[6100 rows x 3 columns]


In [ ]:
#Displaying the training dataset
print(train_texts)

0       "Warning 4668 is an extra one we have enabled ...
1       "Make it possible to optionally declare a func...
2       "It has a new feature to add a route that will...
3       "It's nice to have a notification from a news ...
4       "And please don't always use security risk as ...
                              ...                        
5095    "Now it seems there is zero standard to be cer...
5096    Similarly this idea could be thought through f...
5097           "You see a ad for a pair a shoe you want."
5098    "The current support for Bluetooth LE in Win10...
5099    "I'd like to suggest reverting the change that...
Name: text, Length: 5100, dtype: object


In [ ]:
#Displaying the testing dataset
print(test_texts)

5100    "Please add an API that allows to read the tem...
5101    "The issue we run into is that our desktop app...
5102    "This UI should get loaded on client machine a...
5103    "There need to be a business app settings laun...
5104    "I want to use this control TreeView from this...
                              ...                        
6095    "When the user will press button for next item...
6096    "We end up writing lot of custom Styles in XAM...
6097    "Allow UWP applications to tick a box in the ....
6098     Allow JSON and/or XML to pass data between apps.
6099    "I have added this whole ASN (AS8075 https://i...
Name: text, Length: 1000, dtype: object


In [ ]:
#Displaying the training labels
print(train_labels) 

0       0
1       1
2       0
3       0
4       1
       ..
5095    0
5096    1
5097    0
5098    0
5099    1
Name: label, Length: 5100, dtype: int64


In [ ]:
#Displaying the testing labels
print(test_labels)

5100    1
5101    0
5102    1
5103    1
5104    0
       ..
6095    0
6096    0
6097    1
6098    1
6099    0
Name: label, Length: 1000, dtype: int64


**Task 2: Data Pre-processing (30 Marks)**
**Explain at least 3 steps that you will perform to preprocess the texts before training a classifier.**

Text Preprocessing is the process of cleaning textual data before it is given as input data to the model.  The raw data contains unwanted punctuation marks, numbers and alphabets in upper and lower case etc which can interfere during text analysis.

Data cleaning is defined as a procedure to add  missing data, remove unwanted and irrelevant data and discard duplicate values from the input dataset.

Hence, text pre-processing is done to ensure that all irregularities in the input data is removed.

> I have implemented the following text pre-processing techniques:

**1. Removal of all the Special Characters and Punctuation Marks**

>The special characters do not carry any important information. Therefore, I have  removed all the URLs, symbols etc., from the dataset for data cleaning. Discarding  the unwanted special characters and punctuation marks can help focus on the relevant information.I have used the Regular Expression(re) library in python to remove all the special characters and punctuation marks.

**2.   Removal of all the numbers.**

>The dataset consists of numbers, but the numbers do not contribute to the suggestion detection. Thus, all the digits in the dataset have been discarded.

**3. Case Normalization.**

> Capital letters are used to indicate the begining of a new sentence. But the upper case letters can add noise while training the data. 
lower() is used to convert all the text into lowercase. This improves simplicity and maintains consistency while training the model.

**4. Stopword Removal.**

> Stopwords are the most commonly occured words in English language. Stopwords mostly include "in", "is" etc,. among others. The stopwords have high frequency of occurences but do not contribute to Suggestion Detection. Therefore, it is important to remove them in order to condserve computing time for large volumes of data.

**5. Removal of Spaces**

> The inconsistent blank spaces are commonly found in any dataset and can lead to genearting inaccurate results during data analysis. Thus, it is imperative to remove any blank spaces from the dataset. 




In the code cell below, write an implementation of the steps you defined above. You are free to use a library such as nltk or sklearn for this task.

In [ ]:
#Importing the libraries
import nltk #Importing the Natural Language Toolkit
import re #Importing the Regular Expressions
from nltk.stem import WordNetLemmatizer #Importing the WordNetLemmatizer 
from nltk.stem import PorterStemmer #Importing the PorterStemmer
from nltk.corpus import stopwords #Importing the stopwords
from nltk.tokenize import word_tokenize #Importing the word_tokenize

In [ ]:
# Defining a method to clean the text
def clean_method(text):
 new=[]
 for i in text:
   
# Removes all special characters and numericals leaving the alphabets
   i = re.sub('[^A-Za-z]+', " ", i)
   i= i.lower()#Converting all the letters into lower case
   i = re.sub("s+"," ",i) #Rewmoval of blankspaces
stopwords = ["the", "man", "a", "about","above","after","again","against","ain","all","am","an","and","any","are","aren","aren't","as","at","be","because","been","before","being","below","between","both","but","by","can","couldn","couldn't","d","did","didn","didn't","do","does","doesn","doesn't","doing","don","don't","down","during","each","few","for","from","further","had","hadn","hadn't","has","hasn","hasn't","have","haven","haven't","having","he","her","here","hers","herself","him","himself","his","how","i","if","in","into","is","isn","isn't","it","it's","its","itself","just","ll","m","ma","me","mightn","mightn't","more","most","mustn","mustn't","my","myself","needn","needn't","no","nor","not","now","o","of","off","on","once","only","or","other","our","ours","ourselves","out","over","own","re","s","same","shan","shan't","she","she's","should","should've","shouldn","shouldn't","so","some","such","t","than","that","that'll","the","their","theirs","them","themselves","then","there","these","they","this","those","through","to","too","under","until","up","ve","very","was","wasn","wasn't","we","were","weren","weren't","what","when","where","which","while","who","whom","why","will","with","won","won't","wouldn","wouldn't","y","you","you'd","you'll","you're","you've","your","yours","yourself","yourselves","could","he'd","he'll","he's","here's","how's","i'd","i'll","i'm","i've","let's","ought","she'd","she'll","that's","there's","they'd","they'll","they're","they've","we'd","we'll","we're","we've","what's","when's","where's","who's","why's","would","able","abst","accordance","according","accordingly","across","act","actually","added","adj","affected","affecting","affects","afterwards","ah","almost","alone","along","already","also","although","always","among","amongst","announce","another","anybody","anyhow","anymore","anyone","anything","anyway","anyways","anywhere","apparently","approximately","arent","arise","around","aside","ask","asking","auth","available","away","awfully","b","back","became","become","becomes","becoming","beforehand","begin","beginning","beginnings","begins","behind","believe","beside","besides","beyond","biol","brief","briefly","c","ca","came","cannot","can't","cause","causes","certain","certainly","co","com","come","comes","contain","containing","contains","couldnt","date","different","done","downwards","due","e","ed","edu","effect","eg","eight","eighty","either","else","elsewhere","end","ending","enough","especially","et","etc","even","ever","every","everybody","everyone","everything","everywhere","ex","except","f","far","ff","fifth","first","five","fix","followed","following","follows","former","formerly","forth","found","four","furthermore","g","gave","get","gets","getting","give","given","gives","giving","go","goes","gone","got","gotten","h","happens","hardly","hed","hence","hereafter","hereby","herein","heres","hereupon","hes","hi","hid","hither","home","howbeit","however","hundred","id","ie","im","immediate","immediately","importance","important","inc","indeed","index","information","instead","invention","inward","itd","it'll","j","k","keep","keeps","kept","kg","km","know","known","knows","l","largely","last","lately","later","latter","latterly","least","less","lest","let","lets","like","liked","likely","line","little","'ll","look","looking","looks","ltd","made","mainly","make","makes","many","may","maybe","mean","means","meantime","meanwhile","merely","mg","might","million","miss","ml","moreover","mostly","mr","mrs","much","mug","must","n","na","name","namely","nay","nd","near","nearly","necessarily","necessary","need","needs","neither","never","nevertheless","new","next","nine","ninety","nobody","non","none","nonetheless","noone","normally","nos","noted","nothing","nowhere","obtain","obtained","obviously","often","oh","ok","okay","old","omitted","one","ones","onto","ord","others","otherwise","outside","overall","owing","p","page","pages","part","particular","particularly","past","per","perhaps","placed","please","plus","poorly","possible","possibly","potentially","pp","predominantly","present","previously","primarily","probably","promptly","proud","provides","put","q","que","quickly","quite","qv","r","ran","rather","rd","readily","really","recent","recently","ref","refs","regarding","regardless","regards","related","relatively","research","respectively","resulted","resulting","results","right","run","said","saw","say","saying","says","sec","section","see","seeing","seem","seemed","seeming","seems","seen","self","selves","sent","seven","several","shall","shed","shes","show","showed","shown","showns","shows","significant","significantly","similar","similarly","since","six","slightly","somebody","somehow","someone","somethan","something","sometime","sometimes","somewhat","somewhere","soon","sorry","specifically","specified","specify","specifying","still","stop","strongly","sub","substantially","successfully","sufficiently","suggest","sup","sure","take","taken","taking","tell","tends","th","thank","thanks","thanx","thats","that've","thence","thereafter","thereby","thered","therefore","therein","there'll","thereof","therere","theres","thereto","thereupon","there've","theyd","theyre","think","thou","though","thoughh","thousand","throug","throughout","thru","thus","til","tip","together","took","toward","towards","tried","tries","truly","try","trying","ts","twice","two","u","un","unfortunately","unless","unlike","unlikely","unto","upon","ups","us","use","used","useful","usefully","usefulness","uses","using","usually","v","value","various","'ve","via","viz","vol","vols","vs","w","want","wants","wasnt","way","wed","welcome","went","werent","whatever","what'll","whats","whence","whenever","whereafter","whereas","whereby","wherein","wheres","whereupon","wherever","whether","whim","whither","whod","whoever","whole","who'll","whomever","whos","whose","widely","willing","wish","within","without","wont","words","world","wouldnt","www","x","yes","yet","youd","youre","z","zero","a's","ain't","allow","allows","apart","appear","appreciate","appropriate","associated","best","better","c'mon","c's","cant","changes","clearly","concerning","consequently","consider","considering","corresponding","course","currently","definitely","described","despite","entirely","exactly","example","going","greetings","hello","help","hopefully","ignored","inasmuch","indicate","indicated","indicates","inner","insofar","it'd","keep","keeps","novel","presumably","reasonably","second","secondly","sensible","serious","seriously","sure","t's","third","thorough","thoroughly","three","well","wonder","a", "about", "above", "above", "across", "after", "afterwards", "again", "against", "all", "almost", "alone", "along", "already", "also","although","always","am","among", "amongst", "amoungst", "amount", "an", "and", "another", "any","anyhow","anyone","anything","anyway", "anywhere", "are", "around", "as", "at", "back","be","became", "because","become","becomes", "becoming", "been", "before", "beforehand", "behind", "being", "below", "beside", "besides", "between", "beyond", "bill", "both", "bottom","but", "by", "call", "can", "cannot", "cant", "co", "con", "could", "couldnt", "cry", "de", "describe", "detail", "do", "done", "down", "due", "during", "each", "eg", "eight", "either", "eleven","else", "elsewhere", "empty", "enough", "etc", "even", "ever", "every", "everyone", "everything", "everywhere", "except", "few", "fifteen", "fify", "fill", "find", "fire", "first", "five", "for", "former", "formerly", "forty", "found", "four", "from", "front", "full", "further", "get", "give", "go", "had", "has", "hasnt", "have", "he", "hence", "her", "here", "hereafter", "hereby", "herein", "hereupon", "hers", "herself", "him", "himself", "his", "how", "however", "hundred", "ie", "if", "in", "inc", "indeed", "interest", "into", "is", "it", "its", "itself", "keep", "last", "latter", "latterly", "least", "less", "ltd", "made", "many", "may", "me", "meanwhile", "might", "mill", "mine", "more", "moreover", "most", "mostly", "move", "much", "must", "my", "myself", "name", "namely", "neither", "never", "nevertheless", "next", "nine", "no", "nobody", "none", "noone", "nor", "not", "nothing", "now", "nowhere", "of", "off", "often", "on", "once", "one", "only", "onto", "or", "other", "others", "otherwise", "our", "ours", "ourselves", "out", "over", "own","part", "per", "perhaps", "please", "put", "rather", "re", "same", "see", "seem", "seemed", "seeming", "seems", "serious", "several", "she", "should", "show", "side", "since", "sincere", "six", "sixty", "so", "some", "somehow", "someone", "something", "sometime", "sometimes", "somewhere", "still", "such", "system", "take", "ten", "than", "that", "the", "their", "them", "themselves", "then", "thence", "there", "thereafter", "thereby", "therefore", "therein", "thereupon", "these", "they", "thickv", "thin", "third", "this", "those", "though", "three", "through", "throughout", "thru", "thus", "to", "together", "too", "top", "toward", "towards", "twelve", "twenty", "two", "un", "under", "until", "up", "upon", "us", "very", "via", "was", "we", "well", "were", "what", "whatever", "when", "whence", "whenever", "where", "whereafter", "whereas", "whereby", "wherein", "whereupon", "wherever", "whether", "which", "while", "whither", "who", "whoever", "whole", "whom", "whose", "why", "will", "with", "within", "without", "would", "yet", "you", "your", "yours", "yourself", "yourselves", "the","a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p",
]
#Defining a list of our own stopwords

temp_train = clean_method(train_texts) #Calling the clean method on training data
temp_test = clean_method(test_texts) #Calling the clean method for testing data
print(temp_train)


# Defining a method to remove the stopwords
def rem(text):
  out_ar=[]
  for cem in text:
    temp_list = []
    for w in cem.split():
      if w not in stopwords:
        temp_list.append(w)
    out_ar.append(' '.join(temp_list))
  return out_ar          
temp_train = rem(train_texts) #Calling the stopword removal method on training data
temp_test = rem(test_texts)   #Calling the stopword removal method on training data


None


In [ ]:
 #Displaying the dataype of the temp_train
type(temp_train)

list

In [ ]:
import nltk #Importing the Natural Language Toolkit Library
nltk.download('punkt') #It is a sentence tokenizer that divides the text into a list of sentences

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
print (temp_train)

['"Warning 4668 extra enabled warning level 4."', '"Make optionally declare function OnClick() live tiles, code executed live tiles presed leaving screen."', '"It feature add route displayed street names."', '"It\'s nice notification news read short extract experience - dynamic rich notifications - displayed gesture left screen (from edge swipe W8.1 apps running) - lock screen customizable 5___ more"', '"And security risk excuse allowing people create creative apps Edge."', '"_______________MediaElement_________HTTP__________________Cookie_________________________________________MediaElement___CookieContainer____________________________________"', 'It indexer working properly (which uncommon).', '"I control balance audio I play BackgroundAudioPlayer."', '"(Pages created Wordpress this.)""', '"You select text big screen."', '"Photos, email, internet, games, stability learning curve."', '"(If PInvoke.Kernel32.dll install nuget-package directly UWP-project automatic references nuget-packa

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer #Importing CountVectorizer from sklearn
from sklearn.feature_extraction.text import TfidfTransformer #Importing TfidfVectorizer from sklearn

count_vect = CountVectorizer()  #Countvectorizer is a method to convert text to numerical data.
#It is used to convert a collection of raw documents to a matrix of TF-IDF features. 
X_train_counts = count_vect.fit_transform(temp_train) 
#Transform a count matrix to a normalized tf or tf-idf representation.
tfidf_transformer = TfidfTransformer()
X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)
# Train a Naïve Bayes classifier using the tf-idf scores for words as features.
# Transforming test data in into tf-idf vector
X_test_counts = count_vect.transform(temp_test) 
X_test_tfidf = tfidf_transformer.transform(X_test_counts)

In [ ]:
#Importing GaussianNB from sklearn
from sklearn.naive_bayes import GaussianNB
set1_NB_classifier_tfidf = GaussianNB()
set1_NB_classifier_tfidf.fit(X_train_tfidf.toarray(), train_labels)
# evaluation
predictions = set1_NB_classifier_tfidf.predict(X_test_tfidf.toarray())

def accuracy(labels, predictions):
    """
      Calculate the accuracy score for a given set of predictions and labels.
  
      Args:
       labels (list): A list containing gold standard labels annotated as `0` and `1`.
        predictions (list): A list containing predictions annotated as `0` and `1`.

      Returns:
       float: A floating point value to score the predictions against the labels.
  """

    assert len(labels) == len(predictions)
  
    correct = 0
    for label, prediction in zip(labels, predictions):
        if label == prediction:
            correct += 1 
    score = correct / len(labels)

    return score

# Calculate accuracy score for the classifier using tf-idf features.
accuracy(test_labels, predictions)

0.559

**Task 4: Evaluation Metrics (15 marks)**

**Why is accuracy not the best measure for evaluating a classifier? Describe an evaluation metric**

Accuracy can be defined as the ratio of correct predictions to total number of predictions.
Accuracy is a single metric that can be used to summarise a model.
Accuracy works the best for a Balanced Classification  when there are equal number of samples in each class.
Accuracy cannot be used for a classification task whose distribution of class is skewed or on an imbalanced classification because of the intuition of a ML Practitioner.
ML Practitioners usually thaink that small rates of error and high accuracy scores such as close to 90 or above are ideal to evaluate the performance of model since they usuaaly work on small sized datasets with equal distribution of class.

Therefore, accuracy cannot be implemented on imbalanced class distribution.



**Which might work better than accuracy for a classification task such as suggestion detection.**

There are multiple evaluation metrics such as Accuracy, Precision, Recall nad F1 Score.

**Precision:**
It can be defined as the ratio of True Positives to the sum of True Positives and False Positives.
It does not consider True Negatives into consideration.
It is only best suited for Binary Classification tasks.
Comparison of the models cannot be done because it is not a single metrics.

**Recall:**
It can be defined as the ratio of True Positives to the sum of True Positives and False Negatives.
It does not make use of all the fields in the confusion matrix.
It focuses only on the Positive Class.
Comparison of the models cannot be done because it is not a single metrics.

**F-Score**

F-Score can be defined as the harmonic mean of Precision and Recall.
It is the ratio of product of Precision and Recall to the sum of Precision and Recall multiplied by 2.
The highest vale of F-Score is 1. It balances both Precision and Recall.
Since it is a single metrics it can be used to compare the models.
False Negatives can be minimized by maximizing precision and False Positives can be minimized by maximizing recall. Precision is given more significance.

**F Beta Score**

It prioritises either precision or accuarcy by adding additional weights in the form of Beta values.
The beta value is less than 1 when false positives are minimized whereas beta value is set to greater than 1 when false negatives are minimized. Recall is given more significance.

**Therefore, F-Score and F Beta Score are the most efficient evaluation metrics.**




Edit this cell to write your answer below the line in no more than 150 words.
Delete this line and write your answer here
In the code cell below, write an implementation of the evaluation metric you defined above. You are
free to use a library such as nltk or sklearn for this task, or you can write your own implementation
from scratch.

In [ ]:
def evaluate(labels, predictions):
  '''
  Calculate an evaluation score other than accuracy for a given set of predictions and labels.
  
  Args:
    labels (list): A list containing gold standard labels annotated as `0` and `1`.
    predictions (list): A list containing predictions annotated as `0` and `1`.

  Returns:
    float: A floating point value to score the predictions against the labels.
  '''

  # check that labels and predictions are of same length
  assert len(labels) == len(predictions)

  score = 0.0
  
  #################### EDIT BELOW THIS LINE #########################
  from sklearn.metrics import f1_score
  score = f1_score(test_labels, predictions)


  #################### EDIT ABOVE THIS LINE #########################

  return score
  

  
# Calculate evaluation score based on the metric of your choice
# for the classifier trained in Task 3 using tf-idf features.
evaluate(test_labels, predictions)


0.5190839694656488

In [ ]:
from sklearn.naive_bayes import GaussianNB
from tqdm import tqdm
from sklearn.metrics import precision_recall_fscore_support, classification_report

In [ ]:
# Lets train a Gaussian Naive Bayes clasifier using counts 
NB_classifier_counts = GaussianNB()
NB_classifier_counts.fit(X_train_counts.toarray(), train_labels)

GaussianNB(priors=None, var_smoothing=1e-09)

In [ ]:
# evaluation
import numpy as np
from sklearn.metrics import classification_report
import pandas as pd
#predictions = NB_classifier_counts.predict(X_test_counts.toarray())
preds = NB_classifier_counts.predict(X_test_counts.toarray())
print(classification_report(test_labels, predictions))


              precision    recall  f1-score   support

           0       0.74      0.50      0.59       648
           1       0.42      0.68      0.52       352

    accuracy                           0.56      1000
   macro avg       0.58      0.59      0.56      1000
weighted avg       0.63      0.56      0.57      1000



**Task 5: Feature Engineering (II) - Other features (40 Marks)**
Describe features other than those defined in Task 3 which might improve the performance of your
suggestion detector. If these features require any additional pre-processing steps, then define those
steps as well.
Edit this cell to write your answer below the line in no more than 500 words.
**n-grams**
Language Model is used to predict the probabability of word-occurences.
n-gram can be defined as a sequence of n words.
The value of n for n-gram can start from n= 1, 2, 3,......
If n is set to 1 then it is a **unigram** with a single word sequence like "Suggestions".
If n is set to 2 then it is a **bigram** with a double word sequence like "See you".
If n is set to 3 then it is a **trigram** ith a triple word sequence like "See you soon".
n- grams can be defined as a set of co-occuring words. n-grams play a  vital role in suggestion detection. n-grams based approach is used to find similarity between texts. . The n-grams’ frequencies are calculated and is given as the input to calculate the accuracy and other evaluation metrics.
To get accurate results, we extract word-level n-grams and analyze them. It allows us to find the same phrases between different texts. Thus, we have a bag of n-grams which is used to compute the accuracy.


The Count Vectorizer imported from sklearn is used to convert the collection of  n-grams obtained from pre-processed data  in to matrix consisting the token-counts.

TfidfTransformer is used to conert a count matrix to a normalized tf or tf-idf representation. Tfidf stands for term-frequency times inverse document-frequency. Term Frequency refers to the number of occurences of a word in a document.But if a term frequently occurs in all documents then the term dies not contribute much to the classification of the document. This is known as  inverse document-frequency.

fit_transform () is applied on the Training data to analyse the vocabulary dictionary and returns document-term matrix.

tfidf.transform() is applied on the Testing data into tf-idf vector.
GaussianNB()
NBClassifier.predict() performs Naive Bayes Classification on  on an array of test vectors X_Counts.
NB_classifier.fit() is applied on training data and train labels to train the model using Naive Bayes Classifier. The Classification report is used to compute and display the values of accuracy, precision, recall and f1-score.

The accuracy value using Count Vectorizer is 67% or 0.67. The other evaluation metrics and the classification report computed using Count Vectorizer is displayed below.

The accuracy value using tfidf is 66% or 0.66. The other evaluation metrics and the classification report computed using Count Vectorizer is displayed below.

The initial accuracy of the model before feature engineering was 55.9% or 0.559.
After feature engineetring it has increased to 66% or 0.66 and 67% or 0.67

In the code cell below, write an implementation of the features (and any additional pre-preprocessing
steps) you defined above. You are free to use a library such as nltk or sklearn for this task.
After creating your features, use the training data to train a Naïve Bayes classifier and use the test
set to evaluate its performance using the metric defined in Task 4. You must not use the test set
for training.
To make sure that your code doesn’t take too long to run or use too much memory, you can consider
a time limit of 3 minutes and a memory limit of 12GB for this task.

In [23]:
!curl "https://raw.githubusercontent.com/pasricha/Subtask-A/master/Data.Assignment2.SemEvalTask9SubtaskA.csv" > Data.Assignment2.SemEvalTask9SubtaskA.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  729k  100  729k    0     0  7516k      0 --:--:-- --:--:-- --:--:-- 7516k


In [24]:
#Importing the libraries
import pandas as pd 
import numpy as np
import nltk
import re 
from nltk.stem import WordNetLemmatizer
from nltk.stem import PorterStemmer
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import nltk
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


True

In [25]:
!curl "https://raw.githubusercontent.com/pasricha/Subtask-A/master/Data.Assignment2.SemEvalTask9SubtaskA.csv" > Data.Assignment2.SemEvalTask9SubtaskA.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  729k  100  729k    0     0  4263k      0 --:--:-- --:--:-- --:--:-- 4263k


In [26]:
import numpy as np
import pandas as pd

# Read the CSV file.
df = pd.read_csv('Data.Assignment2.SemEvalTask9SubtaskA.csv', 
                 names=['id', 'text', 'label'],header = 0 )

# Set seed for reproducibility and shuffle the rows.
np.random.seed(888)
df = df.sample(frac=1).reset_index(drop=True)

# Store the data as a list of tuples where the first item is the text
# and the second item is the label.
data = [(text, label) for (idx, text, label) in df.values.tolist()]

# Create training and test sets.
train_texts, train_labels = [], []
test_texts, test_labels = [], []

#################### EDIT BELOW THIS LINE #########################

train_data = data[:5100]
test_data = data[5100:]

train_texts, train_labels =  zip(*train_data)
test_texts, test_labels = zip(*test_data)


#################### EDIT ABOVE THIS LINE #########################

# Check that training set and test set are of the right size.
assert len(test_texts) == len(test_labels) == 1000
assert len(train_texts) == len(train_labels) == 5100

In [27]:
#Definig a method to clean the data
def clean(text):
# Removes all special characters and numericals leaving the alphabets
    text = re.sub('[^A-Za-z]+', ' ', text)
    text = text.lower()
    return text
temp_train= []

# for text in train_texts:
#   temp_train.append(clean(text))
# temp_test= []
# for text in test_texts:
#   temp_test.append(clean(text))


df['Cleaned Text'] = df['text'].apply(clean)
df.head()

id  ...                                       Cleaned Text
0    4191  ...   warning is an extra one we have enabled beyon...
1  978_10  ...   make it possible to optionally declare a func...
2  1274_2  ...   it has a new feature to add a route that will...
3   793_8  ...   it s nice to have a notification from a news ...
4   0_421  ...   and please don t always use security risk as ...

[5 rows x 4 columns]

In [28]:
import nltk #Importing the Natural Language Toolkit
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


True

In [31]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [32]:
from sklearn.feature_extraction.text import CountVectorizer #Importing CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer #Importing TfidfTransformer
from sklearn.naive_bayes import GaussianNB #Importing GaussianNB

count_vect = CountVectorizer(ngram_range=(1, 2)) #Countvectorizer is a method to convert text to numerical data.


X_train_counts = count_vect.fit_transform(train_texts)  
#It is used to convert a collection of raw documents to a matrix of TF-IDF features. 
tfidf_transformer = TfidfTransformer()

X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)

train_features = X_train_counts.toarray() + X_train_tfidf.toarray()


# print(train_features.shape)
# Transforming test data in into tf-idf vector
X_test_counts = count_vect.transform(test_texts) 
X_test_tfidf = tfidf_transformer.transform(X_test_counts)

#test_features = X_test_counts.toarray() + X_test_tfidf.toarray()


In [33]:

from sklearn.naive_bayes import GaussianNB
from tqdm import tqdm
from sklearn.metrics import precision_recall_fscore_support, classification_report, accuracy_score

In [34]:
# Lets train a Gaussian Naive Bayes clasifier using counts 
NB_classifier = GaussianNB()
NB_classifier.fit(X_train_counts.toarray(), train_labels)



GaussianNB(priors=None, var_smoothing=1e-09)

In [35]:
# evaluation
preds = NB_classifier.predict(X_test_counts.toarray())
#preds = NB_classifier.predict(X_test_tfidf.toarray())

print(accuracy_score(test_labels, preds))
print(classification_report(test_labels, preds))

0.674
              precision    recall  f1-score   support

           0       0.75      0.74      0.75       648
           1       0.54      0.56      0.55       352

    accuracy                           0.67      1000
   macro avg       0.65      0.65      0.65      1000
weighted avg       0.68      0.67      0.68      1000



The accuracy is 0.674 or 67.4% using Count Vectorizer.

In [36]:
NB_classifier = GaussianNB()
NB_classifier.fit(X_train_tfidf.toarray(), train_labels)


GaussianNB(priors=None, var_smoothing=1e-09)

In [37]:
# evaluation
#preds = NB_classifier.predict(X_test_counts.toarray())
preds = NB_classifier.predict(X_test_tfidf.toarray())

print(accuracy_score(test_labels, preds))
print(classification_report(test_labels, preds))

0.662
              precision    recall  f1-score   support

           0       0.74      0.75      0.74       648
           1       0.52      0.51      0.51       352

    accuracy                           0.66      1000
   macro avg       0.63      0.63      0.63      1000
weighted avg       0.66      0.66      0.66      1000



The accuracy is 0.662 or 66.2% using tfidf.


**References:**

https://medium.com/analytics-vidhya/performance-metrics-classification-model-69a5546b118c
https://machinelearningmastery.com/failure-of-accuracy-for-imbalanced-class-distributions/
https://machinelearningmastery.com/fbeta-measure-for-machine-learning/https://en.wikipedia.org/wiki/F-score
https://medium.com/@douglaspsteen/beyond-the-f-1-score-a-look-at-the-f-beta-score-3743ac2ef6e3
https://towardsdatascience.com/explaining-precision-vs-recall-to-everyone-295d4848edaf
https://medium.com/analytics-vidhya/performance-metrics-classification-model-69a5546b118c
https://www.analyticsvidhya.com/blog/2021/06/rule-based-sentiment-analysis-in-python/
https://www.analyticsvidhya.com/blog/2019/08/comprehensive-guide-language-model-nlp-python-code/
https://towardsdatascience.com/understanding-word-n-grams-and-n-gram-probability-in-natural-language-processing-9d9eef0fa058